streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java

```diff
@@ -781,4 +775,47 @@ public void shouldNotAllowToAddGlobalStoreWithSourceNameEqualsProcessorName() {
             sameNameForSourceAndProcessor,
             new MockProcessorSupplier());
     }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowIfNameIsNull() {
+        new InternalTopologyBuilder.Source(null, Collections.emptySet(), null);
```

Comment:
We get an NPE with `null` message here. Not sure if this improves the test? Let me know you think. We cannot set a message because we use `Objects.requireNonNull` that trows the NPE.

Reply:
OK, I overlooked that the overload of `requireNonNull` without the message is used in the code under test. However, I think it would be better to use the overload with message, i.e.,
```
public static <T> T requireNonNull(T obj, String message)
```

Reply:
Well. We never specify a message for internal checks. It's not a user facing API but effectively an internal assertion. So not sure if we gain much adding a message and checking for it. It's a fatal error anyway if we hit the NPE.

If we think we should specify messages, we should change our pattern and do this consistently. Thoughts?

Reply:
OK, makes sense. Didn't know about policy of internal checks. Would be good to have it written down somewhere.

Reply:
+1 should have these kinds of things documented somewhere. I've been trying to keep track of public APIs I found confusing/lacking in docs but would be nice to track internal assertions like this as well

Reply:
I hear you. How do you suggest to tackle this?

Personally, I would prefer to use actual assertions, this would make the difference clear. \cc @guozhangwang @bbejeck @vvcephei @abbccdda

Reply:
Do you mean the `assert` keyword in Java? IIUC assertions need to explicitly turned on at execution time. So you need to rely on the user to turn them on.

Reply:
Yes, I mean `assert` keyword. If we turn it on in our tests, it might be sufficient (it's internal checks)? Not sure.

Other suggestions are more than welcome :)

Reply:
My preference would be `requireNonNull` and add the message.  First IMHO it's always better to have some sort of description v.s just a `NullPointerException` and second, there is less chance for error since we need to explicitly enable assertions.   Just my 2 cents.

Reply:
I am with @bbejeck on this. Alternatively, we could specify a separate exception that shows that a contract was violated internally and gives you information to quickly identify the bug. However, I am not sure whether this additional exception would introduce more complexity than benefits.

I would not use an assert and turn it on only in tests. If a contract is violated internally during production, it could prolong the search for the bug, because the error could pop up in places other than the actual origin of the violation.

Reply:
Ack.